web/auth/auth.go

```diff
@@ -177,145 +144,148 @@ func sessionData(session *sessions.Session) string {
+		redirectURL := providerConfig.AuthCodeURL(secret)
+		logger.Debugf("redirecting to AuthURL: %v", redirectURL)
+		http.Redirect(w, r, redirectURL, http.StatusTemporaryRedirect)
 	}
 }
 
 // OAuth2Callback handles the callback from an oauth2 provider.
-func OAuth2Callback(logger *zap.SugaredLogger, db database.Database, scms *Scms) echo.HandlerFunc {
-	return func(c echo.Context) error {
+func OAuth2Callback(logger *zap.SugaredLogger, db database.Database, authConfig *AuthConfig, scms *Scms, secret string) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
 		logger.Debug("OAuth2Callback: started")
-		w := c.Response()
-		r := c.Request()
 
-		qv := r.URL.Query()
-		logger.Debugf("qv: %v", qv)
-		redirect, teacher := extractState(r, State)
-		logger.Debugf("Redirect: %v ; Teacher: %t", redirect, teacher)
+		if r.Method != "GET" {
+			authenticationError(logger, w, fmt.Sprintf("illegal request method: %s", r.Method))
+			return
+		}
 
-		provider, err := gothic.GetProviderName(r)
-		if err != nil {
-			logger.Error("failed to get gothic provider", zap.Error(err))
-			return echo.NewHTTPError(http.StatusBadRequest, err.Error())
+		// Get provider.
+		provider := GetProviderName(r.URL.Path, 3)
+		if provider == "" {
+			authenticationError(logger, w, "incorrect request URL")
+			return
 		}
-		// Add teacher suffix if upgrading scope.
-		if teacher {
-			qv.Set("provider", provider+TeacherSuffix)
-			logger.Debugf("Set('provider') = %v", provider+TeacherSuffix)
+
+		providerConfig, ok := authConfig.providers[provider]
+		if !ok {
+			authenticationError(logger, w, "provider is not enabled")
+			return
 		}
-		r.URL.RawQuery = qv.Encode()
-		logger.Debugf("RawQuery: %v", r.URL.RawQuery)
 
-		// Complete authentication.
-		externalUser, err := gothic.CompleteUserAuth(w, r)
-		if err != nil {
-			logger.Error("failed to complete user authentication", zap.Error(err))
-			return echo.NewHTTPError(http.StatusBadRequest, err.Error())
+		if err := r.ParseForm(); err != nil {
+			authenticationError(logger, w, fmt.Sprintf("failed to parse request form: %s", err))
+			return
 		}
-		logger.Debugf("externalUser: %v", qlog.IndentJson(externalUser))
 
-		remoteID, err := strconv.ParseUint(externalUser.UserID, 10, 64)
+		// Validate redirect state to make sure the request is in fact coming from the oauth provider.
+		callbackSecret := r.FormValue("state")
+		if callbackSecret != secret {
+			authenticationError(logger, w, "incorrect callback secret")
+			return
+		}
+
+		// Exchange code for access token.
+		code := r.FormValue("code")
+		if code == "" {
+			authenticationError(logger, w, "got empty code on callback")
+			return
+		}
+
+		authToken, err := providerConfig.Exchange(context.Background(), code)
 		if err != nil {
-			logger.Error(err.Error())
-			return err
+			authenticationError(logger, w, fmt.Sprintf("failed to exchange access token: %s", err))
+			return
```

Comment:
Code from lines 207-241 could be made into a single method on `AuthConfig`, obviating the need to expose the `config.OAuth` (`providerConfig`).

PS: The method should return an `error` and not use the `authenticationError` helper function.

Reply:
The proposed method should return a string-based `accessToken` instead of `oauth2.Token` since we don't use it.